### PR TITLE
Fixes #37059 - Templates - cloned_from reference

### PR DESCRIPTION
**Pull request by stejskalleos**
Keep reference to the original template in cloned templates.

### CODE REVIEW

File: app/views/templates/_form.html.erb [hunk 1]
```diff
@@ -199,4 +210,5 @@
 
     <%= submit_or_cancel f, false, :cancel_path => template_path_for(@template.class) %>
   </div>
+  <%= f.hidden_field :cloned_from_id %>
```
Comment (ares) [inline app/views/templates/_form.html.erb:213]: What is this good for? Can I change this internal value if I modify this in browser console?

Comment (ares) [reply app/views/templates/_form.html.erb:213]: If this is removed, I guess it should also be removed from the parameters filter.

Comment (stejskalleos) [reply app/views/templates/_form.html.erb:213]: It's probably a relic from ancient times when I needed it for the implementation, and then I went with a different approach.

I'll remove it.

**Comment by ekohl** [comment]
Not looking at the code, but you used SAT- instead of a Redmine prefix

**Review by ekohl — COMMENTED** [review]
What's the behavior on deletions? If you remove a template that was cloned, should it block that removal or set it to `NULL` (cascade doesn't make sense to me).

**Comment by stejskalleos** [comment]
> What's the behavior on deletions? If you remove a template that was cloned, should it block that removal or set it to `NULL` (cascade doesn't make sense to me).

Fixed in migration:
```ruby
add_reference(:templates, :cloned_from, foreign_key: { to_table: :templates, on_delete: :nullify })
```

**Comment by stejskalleos** [comment]
Reopening, the community is interested in it and requested it.

File: app/models/template.rb [hunk 2]
```diff
@@ -25,6 +26,8 @@ class Template < ApplicationRecord
 
   attr_exportable :name, :description, :snippet, :template_inputs, :model => ->(template) { template.class.to_s }
 
+  scoped_search relation: :cloned_from, on: :cloned_from_id, complete_value: false
```
Comment (ShimShtein) [inline app/models/template.rb]: Why do we need to search by template id? I am not even sure the id is visible in the UI.

Comment (stejskalleos) [reply app/models/template.rb]: I guess searching by name would be more user-friendly, but I'm unsure if it is possible with the `scoped_search` gem.

Comment (ShimShtein) [reply app/models/template.rb]: If you can't find a proper way in the [documentation](https://github.com/wvanbergen/scoped_search/wiki/search-definition), you can always use the [`ext_method`](https://github.com/wvanbergen/scoped_search/wiki/search-definition#external-search-methods), like we did [here](https://github.com/theforeman/foreman/blob/8283db898b7e145e59dd376ac0b4cb2d30f7b436/app/models/fact_value.rb#L20)

Comment (stejskalleos) [reply app/models/template.rb]: Hmm let's remove it for now, I'd like to keep the PR as small as possible

**Comment by stejskalleos** [comment]
@ShimShtein can we get this in?

**Review by ShimShtein — APPROVED** [review]

**Pull request merged**